### PR TITLE
Substitute LSTM with Transformer in Embedding Policy

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- Substitute LSTM with Transformer in Embedding Policy
- fixes https://github.com/RasaHQ/research/issues/34
- fixes https://github.com/RasaHQ/rasa/issues/4089

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by Ghostvv** [comment]
I am adding it right now

File: CHANGELOG.rst [hunk 1]
```diff
@@ -40,6 +43,9 @@ Changed
 -------
 - new event broker class: ``SQLProducer``. This event broker is now used when running locally with
   Rasa X
+
+Removed
```
Comment (tabergma) [inline CHANGELOG.rst]: I guess there was a merge conflict. `Removed` should not be here. Can you please go over the file again?

File: docs/core/policies.rst [hunk 2]
```diff
@@ -165,47 +165,27 @@ set the ``random_seed`` attribute of the ``KerasPolicy`` to any integer.
...
-      attention to;
-    - if the similarity between this previous action embedding and
-      current time dialogue embedding is high, overwrite current LSTM
-      state with the one from the time when this action happened;
-    - for each LSTM time step, calculate the similarity between the
+    - concatenate user input (user intent and entities),
+      previous system action, slots and active form
+      for each time step into an input vector
+      to pre-transformer embedding layer;
+    - feed it to tranformer;
```
Comment (tabergma) [inline docs/core/policies.rst]:
```suggestion
    - feed it to transformer;
```

File: rasa/core/policies/embedding_policy.py [hunk 3]
```diff
@@ -295,116 +233,234 @@ def _load_params(self, **kwargs: Dict[Text, Any]) -> None:
...
+                if shuffle:
+                    ids = np.random.permutation(num_labels)
+                else:
+                    ids = range(num_labels)
+
+                for i in ids:
+                    if num_data_cycles[i] > 0 and not skipped[i]:
+                        skipped[i] = True
+                        continue
+                    else:
```
Comment (tabergma) [inline rasa/core/policies/embedding_policy.py]: minor: else not needed

File: rasa/core/policies/embedding_policy.py [hunk 4]
```diff
@@ -1102,240 +959,298 @@ def _linearly_increasing_batch_size(self, epoch: int) -> int:
...
+        batches_per_epoch = 0
+        while True:
+            try:
+                batch_val_loss, batch_val_acc = self.session.run(
+                    [loss, acc], feed_dict={self._is_training: False}
+                )
+                batches_per_epoch += 1
+                ep_val_loss += batch_val_loss
+                ep_val_acc += batch_val_acc
+            except tf.errors.OutOfRangeError:
```
Comment (tabergma) [inline rasa/core/policies/embedding_policy.py]: That seems kind of hacky. Can you find a better condition for interrupting the while loop?

**Comment by dakshvar22** [comment]
@Ghostvv  Reviewed the TF part comprising of batching, train graph construction. Looks good. One small comment - 
Do you think it would be useful to have a registry for dynamic components like loss function and batching strategy. Very similar to how Keras does it so that adding new types of such components is easy. Also, no if statements within the code as the registry could be a dictionary of key as component string and value as the function.

File: rasa/core/policies/embedding_policy.py [hunk 5]
```diff
@@ -1102,240 +959,298 @@ def _linearly_increasing_batch_size(self, epoch: int) -> int:
...
+            tf.expand_dims(t, 0)
+            for name, t in self.attention_weights.items()
+            if name.endswith("multihead_attention/dot_product_attention")
+        ]
 
-    def _train_tf(
-        self, session_data: SessionData, loss: tf.Tensor, mask: tf.Tensor
+        if attention:
+            return tf.concat(attention, 0)
+        else:
```
Comment (tabergma) [inline rasa/core/policies/embedding_policy.py]: not needed

File: rasa/core/policies/embedding_policy.py [hunk 6]
```diff
@@ -1102,240 +959,298 @@ def _linearly_increasing_batch_size(self, epoch: int) -> int:
...
+                            self._train_op, feed_dict={self._is_training: True}
+                        )
 
-            batch_loss_scales = self._scale_loss_by_count_actions(
-                session_data.X,
-                session_data.slots,
-                session_data.previous_actions,
-                session_data.actions_for_Y,
-            )
+                    except tf.errors.OutOfRangeError:
```
Comment (tabergma) [inline rasa/core/policies/embedding_policy.py]: same as above - is there a better condition?

File: rasa/core/policies/embedding_policy.py (hunk 4 above)
Comment (Ghostvv) [reply rasa/core/policies/embedding_policy.py]: this is how they recommend to do it in tf docs: https://www.tensorflow.org/guide/datasets#consuming_values_from_an_iterator

File: rasa/core/policies/embedding_policy.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/core/policies/embedding_policy.py]: I think we need it because `skipped[i]` might be `True` from previous cycle

**Comment by Ghostvv** [comment]
@tmbo currently I make `tensor2tensor` library as requirement, is it ok?

**Comment by sibbsnb** [comment]
Did you see any difference in performance with this change? Any comparison stats on this compared to REDP?

**Comment by Ghostvv** [comment]
@sibbsnb we did experiments on the same dataset as in the paper, this one performs the same, but trains significantly faster

**Comment by sibbsnb** [comment]
Ok thanks. Yeah training is indeed slow for
Redp.

**Comment by Ghostvv** [comment]
Next stage PR is here: https://github.com/RasaHQ/rasa/pull/4218
